### ISSUE TITLE
Use attributes to mark properties that should/should not be injected

### ISSUE THREAD

**Issue by twsouthwick** [autofac/Autofac#620]
I'm using AutoFac to resolve properties in a PowerShell cmdlet, but find that the extension method ResolutionExtensions.InjectUnsetProperties does not provide the behavior I need.  I want to mark certain properties as 'Do Not Inject' as they are optional parameters for cmdlet.  An easy work around would be to add an attribute that causes the resolution logic to skip that specific property.

**Comment by dazinator** [comment]
+1

**Comment by MpDzik** [comment]
+1

**Comment by cbfrank** [comment]
+1

**Comment by tillig** [comment]
@twsouthwick Just to close the loop - you submitted PR #622 (which looks good)... would that work _instead_ of specifically adding attributes?

I'm also thinking about custom attributes vs. reuse of existing attributes. What if we added support in the Autofac.Mef library to support `[Import]` in combination with PR #622 - so you would mark the properties you want Autofac to inject with an `[Import]` attribute. That would avoid tying Autofac core to a custom attribute and would look something like

`builder.RegisterType<Foo>().InjectImportProperties()`

which would then call the delegate filter in PR #622 to attach whitelisting to properties marked with `[Import]`.

What do you think?

**Comment by cbfrank** [comment]
Can autofac support to register types with attribute either? for example:
[Reg(AsType, Name/Key...)]
public class Service:IService{
so that autofac can register all type defined in assmebly automatically

Thanks

**Comment by tillig** [comment]
No, there is not currently attribute-based type registration.

**Comment by cbfrank** [comment]
Thanks, will autofac support this kind of registration in the future?

**Comment by tillig** [comment]
Possibly through an extension or community contribution, but generally core Autofac has tried to be more unobtrusive - register by assembly scanning and _your own_ custom attributes rather than littering your code with Autofac-specific attributes.

**Comment by cbfrank** [comment]
I see. Thank you!

**Comment by alexmg** [comment]
Perhaps we could add a `FindPropertiesWith` method similar to the `FindConstructorsWith` method, that allows the user to provide there own `IPropertyFinder` implementation similar to to `IConstructorFinder`. This would only be called with a list of properties that had already been checked as settable, and we would ship a standard implementation that could be replaced. That way we aren't limiting how the properties are discovered, and the user could choose attributes or any other convention they liked, and keep the logic encapsulated in their own classes. Food for thought.

**Comment by twsouthwick** [comment]
It's great to see discussion on this going now :) @tillig I created this issue then the PR and realized during creating it that attributes would be confining and impose one out of many possible ways to decide where to inject. I'm not set on the PR I submitted, but more as an idea for implementation. I was not familiar with the `FindConstructorsWith` method that @alexmg mentioned, but after looking at it, seems like it would be a good way to solve this issue (and follow the current convention). I can update my PR to prototype that idea if wanted.
